### PR TITLE
EndUserSpanProcessor integration

### LINKED ISSUE #34526: OTEL - Automatic ENDUSER_ID and ENDUSER_ROLE filling
### Description

Hi guys,
as discussed here #33349 it would be nice for Quarkus OpenTelemetry Extension to automatically fill enduser.id and enduser.role when possible, maybe there could be a configuration to enable this behavior.

I am doing this in my projects with a custom SpanProcessor, but maybe this could be of interest for others too (without reinventing the wheel).

What do you think ?

### Implementation ideas

```java
package com.infolabtech.commons;

import jakarta.enterprise.inject.spi.CDI;
import java.security.Principal;
import java.util.Collection;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;
import org.apache.commons.collections4.CollectionUtils;
import org.apache.commons.collections4.IterableUtils;
import org.apache.commons.lang3.StringUtils;
import org.eclipse.microprofile.jwt.JsonWebToken;

public interface Authenticated {

    default Collection<Principal> getPrincipals() {
        return CollectionUtils.select(
                IterableUtils.toList(
                        CDI.current().select(
                                Principal.class
                        )
                ),
                principal -> StringUtils.isNotEmpty(
                        principal.getName()
                )
        );
    }

    default Set<String> getUsernames() {
        return getPrincipals().stream().map(
                Principal::getName
        ).collect(
                Collectors.toSet()
        );
    }

    default Optional<Principal> getPrincipal() {
        return getPrincipals().stream().findFirst();
    }

    default Optional<String> getUsername() {
        return getPrincipal().map(
                Principal::getName
        );
    }

    default Optional<Collection<String>> getGroups() {
        return getPrincipal().filter(
                JsonWebToken.class::isInstance
        ).map(
                JsonWebToken.class::cast
        ).map(
                JsonWebToken::getGroups
        );
    }

}
```

```java
package com.infolabtech.commons.tracing;

import com.infolabtech.commons.Authenticated;
import io.opentelemetry.api.common.Attributes;
import io.opentelemetry.context.Context;
import io.opentelemetry.sdk.trace.ReadWriteSpan;
import io.opentelemetry.sdk.trace.ReadableSpan;
import io.opentelemetry.sdk.trace.SpanProcessor;
import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.enterprise.context.control.ActivateRequestContext;
import jakarta.inject.Inject;
import org.apache.commons.collections4.CollectionUtils;
import org.eclipse.microprofile.context.ManagedExecutor;

@ApplicationScoped
public class EndUserSpanProcessor implements SpanProcessor, Authenticated {

    @Inject
    protected ManagedExecutor managedExecutor;

    @Override
    @ActivateRequestContext
    public void onStart(Context parentContext, ReadWriteSpan span) {
        managedExecutor.execute(
                () -> getPrincipal().ifPresent(
                        principal -> span.setAllAttributes(
                                Attributes.of(
                                        SemanticAttributes.ENDUSER_ID,
                                        principal.getName(),
                                        SemanticAttributes.ENDUSER_ROLE,
                                        CollectionUtils.emptyIfNull(
                                                principal.getGroups()
                                        ).toString()
                                )
                        )
                )
        );
    }

    @Override
    public boolean isStartRequired() {
        return Boolean.TRUE;
    }

    @Override
    public void onEnd(ReadableSpan span) {
    }

    @Override
    public boolean isEndRequired() {
        return Boolean.FALSE;
    }

}
```

### PR DESCRIPTION
**Pull request by amoscatelli**
Closes #34526

### CODE REVIEW

**Review by brunobat — CHANGES_REQUESTED** [review]
Thanks for the PR  @amoscatelli.
I think we should not add a SpanProcessor and include a ManagedExecutor just for this purpose. 
I was expecting to instrument quarkus-jwt and get that information from there. 
I wonder if  @sberyozkin can help with that?
We also need tests.

**Comment by amoscatelli** [comment]
> Thanks for the PR @amoscatelli. I think we should not add a SpanProcessor and include a ManagedExecutor just for this purpose. I was expecting to instrument quarkus-jwt and get that information from there. I wonder if @sberyozkin can help with that? We also need tests.

I think there is a misunderstanding.
Even if you inject JsonWebToken principal, you still would require a ManagedExecutor to access it.
That was my starting point, but since the beginning vertx complained about being blocked, so I had to introduce a ManagedExecutor. Probably that's because the spanprocessor, accessing the JsonWebToken, transitively access the request and so it requires RequestScope and this operation becomes blocking.

Then I switched to SecurityIdentity, hoping, among the other things, to workaroud the ManagedExecutor thing, but it is still required.
Anyway I think that using SecurityIdentity is still an upgrade, because it is agnostic of the kind of authentication/security extension you are going to use.
There is no reason to couple this with jwt ...

Of course as I said from the beginning I would like to avoid both the ManagedExecutor and the requestscope.
What do you suggest ?

**Comment by sberyozkin** [comment]
LGTM, indeed, should not be tied up with the specific type of credentials or authentication mechanism.
It only should not be enabled by default IMHO which what this PR enforces

**Review by sberyozkin — APPROVED** [review]

**Comment by sberyozkin** [comment]
@amoscatelli Can you please fix the conflict ?

**Review by brunobat — CHANGES_REQUESTED** [review]
Thanks @sberyozkin.
@amoscatelli can you please add tests?

**Comment by amoscatelli** [comment]
> LGTM, indeed, should not be tied up with the specific type of credentials or authentication mechanism. It only should not be enabled by default IMHO which what this PR enforces

Thank you.
Anyway this is not enabled by default.
Default value is false for the new configuration.

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/build/EndUserSpanProcessorConfig.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+    /**
+     * Enable tracing with OpenTelemetry.
+     * <p>
+     * This property is not available in the Open Telemetry SDK. It's Quarkus specific.
+     * <p>
+     * Support for tracing will be enabled if OpenTelemetry support is enabled
+     * and either this value is true, or this value is unset.
+     */
+    @WithDefault("false")
```
Comment (amoscatelli) [inline extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/build/EndUserSpanProcessorConfig.java:22]: @sberyozkin see here ?

File: integration-tests/opentelemetry/src/main/resources/application.properties [hunk 2]
```diff
@@ -6,5 +6,7 @@ quarkus.application.version=999-SNAPSHOT
 quarkus.otel.bsp.schedule.delay=100
 quarkus.otel.bsp.export.timeout=5s
 
+quarkus.otel.traces.eusp.enabled=true
```
Comment (brunobat) [inline integration-tests/opentelemetry/src/main/resources/application.properties:9]: We also need a disabled test.
Probably a configuration profile will work best.

Comment (amoscatelli) [reply integration-tests/opentelemetry/src/main/resources/application.properties]: Done, I added a profile where eusp is enabled and that's used only for my tests.

**Comment by amoscatelli** [comment]
> Thanks @sberyozkin. @amoscatelli can you please add tests?

Done !

**Comment by amoscatelli** [comment]
@brunobat please tell me if there is something more I can do because this is requiring a rebase after rebase :-)

File: integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/EndUserTest.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import io.quarkus.it.opentelemetry.util.EndUserResource;
+import io.quarkus.test.common.http.TestHTTPEndpoint;
+import io.quarkus.test.junit.QuarkusTest;
+import io.quarkus.test.junit.TestProfile;
+import io.quarkus.test.security.TestSecurity;
+
+@QuarkusTest
+@TestHTTPEndpoint(EndUserResource.class)
+@TestProfile(EndUserProfile.class)
+public class EndUserTest {
```
Comment (brunobat) [inline integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/EndUserTest.java:30]: Could you duplicate this test with the default, disabled config and assert that the user data is null?
```
assertNull(attributes.get(SemanticAttributes.ENDUSER_ID));
assertNull(attributes.get(SemanticAttributes.ENDUSER_ROLE));
```

Comment (amoscatelli) [reply integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/EndUserTest.java:30]: Yes, give me a minute, I missed this comment of yours ! sorry for rerequesting the review :-)

Comment (amoscatelli) [reply integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/EndUserTest.java]: Done !

Comment (brunobat) [reply integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/EndUserTest.java]: No worries

**Review by brunobat — APPROVED** [review]
This looks great!
Thanks @amoscatelli

**Comment by brunobat** [comment]
@amoscatelli please build the extension and the IT test before pushing. The local build will auto format the code.

**Comment by amoscatelli** [comment]
@brunobat I think this is done, can this be merged ?

**Comment by amoscatelli** [comment]
No sorry but I have Linkedin !
You can post on Twitter and tag me there, is this possible ?

**Comment by brunobat** [comment]
sure

**Pull request merged**